ledger/sandbox-on-x/src/classic/scala/com/daml/platform/ledger/SandboxServer.scala

```diff
@@ -104,25 +117,65 @@ final class SandboxServer(
       .map(path => Future(Files.write(path, Seq(port.toString).asJava)).map(_ => ()))
       .getOrElse(Future.unit)
 
-  private def loadPackages(writeService: WriteService)(implicit
-      executionContext: ExecutionContext
-  ): AbstractResourceOwner[ResourceContext, List[Unit]] =
-    ResourceOwner.forFuture(() =>
-      Future.sequence(
-        config.damlPackages.map { file =>
-          val submissionId = Ref.SubmissionId.assertFromString(UUID.randomUUID().toString)
-          for {
-            dar <- Future.fromTry(DarParser.readArchiveFromFile(file).toTry)
-            _ <- writeService
-              .uploadPackages(submissionId, dar.all, None)(
-                LoggingContext.ForTesting,
-                NoOpTelemetryContext,
+  private def loadPackages(writeService: WriteService, indexService: IndexService)(implicit
+      executionContext: ExecutionContext,
+      system: ActorSystem,
+      loggingContext: LoggingContext,
+  ): AbstractResourceOwner[ResourceContext, Unit] =
+    ResourceOwner.forFuture(() => {
+      val packageUploadTimeout = 10.seconds
+
+      val packageSubmissionsTrackerMap = config.damlPackages.map { file =>
+        val uploadCompletionPromise = Promise[Unit]()
+        system.scheduler.scheduleOnce(packageUploadTimeout) {
+          // Package upload timeout, meant to allow fail-fast for testing
+          // TODO Remove once in-memory backend (e.g. H2) is deemed stable
+          uploadCompletionPromise.tryFailure(
+            new RuntimeException(s"Package upload timeout after $packageUploadTimeout")
+          )
+          ()
+        }
+        UUID.randomUUID().toString -> (file, uploadCompletionPromise)
+      }.toMap
+
+      val (killSwitch, packageEntriesStreamDone) = indexService
+        .packageEntries(None)
+        .viaMat(KillSwitches.single)(Keep.right[NotUsed, UniqueKillSwitch])
+        .toMat(Sink.foreach {
+          case PackageEntry.PackageUploadAccepted(submissionId, _) =>
+            packageSubmissionsTrackerMap(submissionId)._2.complete(Success(()))
+            ()
+          case PackageEntry.PackageUploadRejected(submissionId, _, reason) =>
+            packageSubmissionsTrackerMap(submissionId)._2
```

Comment:
would be safer to use `.get`
```
packageSubmissionsTrackerMap.get(submissionId).foreach { case (_, promise) =>
  promise.complete(Success(()))
}
```

Reply:
The `submissionId` must be there. If not, there's a bug that should be caught, not hidden, no?

Reply:
Could it be that another user/test could upload packages at the same time which you will get notified about here?

Reply:
But fine, I'll add more verbose handling.

Reply:
Would it be more helpful to have explicit handling of this situation rather than having `None.get` exception without any context?

Reply:
Added